File: DynamicProgramming/376-wiggle-subsequence.py
class Solution:
  def wiggleMaxLength(self, nums: List[int]) -> int:
    n = len(nums)

    dp = [[0] * n for _ in range(2)]

    dp[0][0] = dp[1][0] = 1
    ans = 1
    for i in range(n):
      for j in range(0, i):
        if nums[j] < nums[i]:
          dp[0][i] = max(dp[0][i], dp[1][j] + 1)
        if nums[j] > nums[i]:
          dp[1][i] = max(dp[1][i], dp[0][j] + 1)

        ans = max(dp[0][i], dp[1][i], ans)

    return ans

File: greedy/376-wiggle-subsequence.cpp
/******************************************
 * author      : utoppia
 * description : solution for leetcode
 * updated at  : 2024-11-18 21:33:05
 * filename    : 376-wiggle-subsequence.cpp
 * language    : cpp
 * editor      : Emacs 27.1
 *****************************************/
#include <algorithm>
#include <bits/stdc++.h>
#include <ios>
#include <iostream>
#include <set>
#include <utility>
#include <vector>

#define vall(v) v.begin(), v.end()
#define pir pair<int, int>

#define pb push_back
#define eb emplace_back
#define fi first
#define snd second
#define sz size

#define ll long long
#define nMax 200005

#define MOD 998244353
using namespace std;

class Solution {
public:
  int wiggleMaxLength(vector<int> &nums) {
    if (nums.size() <= 1)
      return nums.size();
    int cur = 0;
    int pre = 0;
    int ans = 1;
    for (int i = 0; i < nums.size() - 1; ++i) {
      cur = nums[i + 1] - nums[i];
      if (pre <= 0 && cur > 0 || pre >= 0 && cur < 0) {
        ans++;
        pre = cur;
      }
    }
    return ans;
  }
};
